pkg/encryption/encryption.go

```diff
@@ -0,0 +1,227 @@
+		defaultKeyRotationPeriod: c.DefaultKeyRotationPeriod,
+		ctrlClient:               c.CtrlClient,
+		logger:                   c.Logger,
+	}
+
+	return s, nil
+}
+
+func (s *Service) Reconcile() error {
+	ctx := context.TODO()
+
+	var encryptionProviderSecret v1.Secret
+
+	err := s.ctrlClient.Get(ctx, ctrlclient.ObjectKey{
+		Name:      key.SecretName(s.cluster.ClusterName),
+		Namespace: s.cluster.Namespace,
+	}, &encryptionProviderSecret)
+
+	if apierrors.IsNotFound(err) {
+		// create new encryption secret
+		err := s.createNewEncryptionProviderSecret(ctx)
+		if err != nil {
+			s.logger.Error(err, "failed to get encryption provider config secret for cluster")
+			return err
+		}
+	} else if err != nil {
+		s.logger.Error(err, "failed to get encryption provider config secret for cluster")
+		return err
+	} else {
+		// config already exists, check for key rotation
+		err = s.keyRotation(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) Delete() error {
+	ctx := context.TODO()
+	encryptionProviderSecret := v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      key.SecretName(s.cluster.ClusterName),
+			Namespace: s.cluster.Namespace,
+		}}
+
+	err := s.ctrlClient.Delete(ctx, &encryptionProviderSecret)
+	if err != nil {
+		s.logger.Error(err, "failed to delete encryption provider config secret for cluster")
+		return err
+	}
+	return nil
+}
+
+func (s *Service) createNewEncryptionProviderSecret(ctx context.Context) error {
+	// check if there is old encryption config that we can use for migration
+	oldEncryptionSecretName := fmt.Sprintf("%s-encryption", s.cluster.ClusterName)
+
+	var oldEncryptionSecret v1.Secret
+	err := s.ctrlClient.Get(ctx, ctrlclient.ObjectKey{
+		Name:      oldEncryptionSecretName,
+		Namespace: s.cluster.Namespace,
+	},
+		&oldEncryptionSecret)
+
+	var encryptionKey string
+	var encryptionProviderType string
+
+	if apierrors.IsNotFound(err) {
+		// no old key found, lets generate a new one
+		newKey, err := newRandomKey(Poly1305KeyLength)
+		if err != nil {
+			s.logger.Error(err, "failed to generate new random key for encryption")
+			return err
+		}
+		encryptionKey = newKey
+		encryptionProviderType = DefaultEncryptionProviderType
+
+	} else if err != nil {
+		s.logger.Error(err, "failed to get old encryption provider key secret")
+		return err
+	} else {
+		// there is an old encryption key so lets reuse it to avoid breaking cluster
+		if k, ok := oldEncryptionSecret.Data["encryption"]; !ok {
+			s.logger.Error(err, "failed to get encryption key from secret")
+			return err
+		} else {
+			encryptionKey = string(k)
+			// the old encryption key are using old less secure type
+			// https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#providers
+			encryptionProviderType = LegacyEncryptionProviderType
+		}
+	}
+
+	secretData := map[string]string{}
+	// render template for the config
+	{
+		templateData := Data{
+			Keys: []Key{
+				{
+					Name:          "key1",
```

Comment:
Does this name hold any special meaning?

Reply:
I can't say i understand the reason but they use this in the examples and value increment  by key index
https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/